src/languageTools/LanguageClient/LanguageClient.js

```diff
@@ -0,0 +1,232 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+/*global exports, Promise, LanguageClientInfo */
+/*eslint no-console: 0*/
+/*eslint strict: ["error", "global"]*/
+/*eslint max-len: ["error", { "code": 200 }]*/
+"use strict";
+
+var ProtocolAdapter = require("./ProtocolAdapter"),
+    ServerUtils = require("./ServerUtils"),
+    Connection = require("./Connection"),
+    NodeToBracketsInterface = require("./NodeToBracketsInterface").NodeToBracketsInterface,
+    ToolingInfo = LanguageClientInfo.toolingInfo,
+    MESSAGE_TYPE = {
+        BRACKETS: "brackets",
+        SERVER: "server"
+    };
+
+function validateHandler(handler) {
+    var retval = false;
+
+    if (handler && typeof handler === "function") {
+        retval = true;
+    } else {
+        console.warn("Handler validation failed. Handler should be of type 'function'. Provided handler is of type :", typeof handler);
+    }
+
+    return retval;
+}
+
+function LanguageClient(clientName, domainManager, options) {
+    this._clientName = clientName;
+    this._bracketsInterface = null;
+    this._notifyBrackets = null;
+    this._requestBrackets = null;
+    this._connection = null,
+    this._startUpParams = null, //_projectRoot, capabilties, workspaceFolders etc.
+    this._initialized = false,
+    this._onRequestHandler = {};
+    this._onNotificationHandlers = {};
+    this._options = options || null;
+
+
+    this._init(domainManager);
+}
+
+
+LanguageClient.prototype._createConnection = function () {
+    if (!this._options || !this._options.serverOptions) {
+        return Promise.reject("No valid options provided for client :", this._clientName);
+    }
+
+    var restartLanguageClient = this.start.bind(this),
+        stopLanguageClient = this.stop.bind(this);
+
+    var serverOptions = this._options.serverOptions;
+    return ServerUtils.startServerAndGetConnectionArgs(serverOptions)
+        .then(function (connectionArgs) {
+            return Connection.createConnection(connectionArgs.reader, connectionArgs.writer, restartLanguageClient, stopLanguageClient);
+        }).catch(function (err) {
+            console.error("Couldn't establish connection", err);
+        });
+};
+
+LanguageClient.prototype.setOptions = function (options) {
+    if (options && typeof options === "object") {
+        this._options = options;
+    } else {
+        console.error("Invalid options provided for client :", this._clientName);
+    }
+};
+
+LanguageClient.prototype.start = function (params) {
+    var self = this;
+
+    //Check to see if a connection to a language server already exists.
+    if (self._connection) {
```

Comment:
I call it "restart" because the function will be invoked when the connection gets an onClose event. In that case, we'll try to automatically restart the server and reestablish the connection internally.

And it's not the function which restarts the server on workspaceChange. That function is in LanguageClientWrapper.js

You can see its usage in DefaultEventHandlers.js in handleProjectOpen.